typescript/src/modelcontextprotocol/essentials.ts

```diff
@@ -23,35 +29,159 @@ class CommercetoolsAgentEssentials extends McpServer {
+  }
+
+  private initializeTools(context?: Context): void {
+    const filteredTools = this.getFilteredTools();
+    const filteredToolsLength = filteredTools.length;
+    const dynamicToolLoadingThreshold =
+      context?.dynamicToolLoadingThreshold ?? DYNAMIC_TOOL_LOADING_THRESHOLD;
+    const shouldReturnAllTools =
+      filteredTools.length <= dynamicToolLoadingThreshold;
+
+    if (shouldReturnAllTools) {
+      this.registerAllTools(filteredTools);
+    } else {
+      console.error(
+        `Filtered tools (${filteredToolsLength}) > ${DYNAMIC_TOOL_LOADING_THRESHOLD} - Using resource based tool system`
+      );
+
+      this.registerResourceBasedToolSystem(filteredTools);
+    }
+  }
+
+  private getFilteredTools() {
+    return contextToTools(this._processedConfiguration.context).filter((tool) =>
+      isToolAllowed(tool, this._processedConfiguration)
     );
+  }
+
+  private registerAllTools(filteredTools: Tool[]): void {
     filteredTools.forEach((tool) => {
-      this.tool(
-        tool.method,
-        tool.description,
-        tool.parameters.shape,
-        async (arg: any) => {
-          const result = await this._commercetools.run(tool.method, arg);
-          return {
-            content: [
-              {
-                type: 'text' as const,
-                text: String(result),
-              },
-            ],
-          };
+      this.registerSingleTool(tool);
+    });
+  }
+
+  private registerResourceBasedToolSystem(filteredTools: Tool[]): void {
+    const {listAvailableTools, injectTools, executeTool} =
+      contextToToolsResourceBasedToolSystem();
+
+    this.registerSingleTool(listAvailableTools);
+    this.registerInjectToolsTool(injectTools, filteredTools);
+    this.registerExecuteTool(executeTool);
+    // Bulk tool is not a resource based tool, so we need to register it separately
+    contextToBulkTools(this._processedConfiguration.context).forEach((tool) => {
+      this.registerSingleTool(tool);
+    });
+  }
+
+  private registerSingleTool(tool: Tool): void {
+    this.tool(
+      tool.method,
+      tool.description,
+      tool.parameters.shape,
+      async (args: Record<string, unknown>) => {
+        const result = await this._commercetools.run(tool.method, args);
+        return this.createToolResponse(result);
+      }
+    );
+  }
+
+  private registerInjectToolsTool(
+    injectTools: Tool,
+    filteredTools: Tool[]
+  ): void {
+    type ToolShape = z.infer<typeof injectTools.parameters.shape>;
+
+    this.tool(
+      injectTools.method,
+      injectTools.description,
+      injectTools.parameters.shape,
+      async (arg: ToolShape) => {
+        const toolsToInject = filteredTools.filter((tool) =>
+          arg.toolMethods.includes(tool.method)
+        );
+
+        toolsToInject.forEach((tool) => {
+          this.registerSingleTool(tool);
+        });
+
+        const injectedTools = await this.formatInjectedTools(toolsToInject);
```

Comment:
I believe the `formatInjectionTools` is not async. We shouldn't be awaiting it here.